data-analytics/cloud-composer-examples/composer_http_post_example/spark_avg_speed.py

```diff
@@ -0,0 +1,144 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This is script defines a PySpark Job to enhance avro files exported from the BigQuery Public Dataset
+nyc-tlc:yellow.trips to include an additional average speed column. This is a quite simple spark job
+which is merely a placeholder to demonstrate Cloud Composer as a way to automate spinning up a
+Dataproc cluster to run a spark job and tear it down once the job completes.
+"""
+
+import argparse
+import datetime
+import json
+import sys
+
+from pyspark import SparkConf, SparkContext
+
+class AverageSpeedEnhancer:
+    def __init__(self):
+        """This Class serves as a namespace for the business logic function"""
+        self.output_schema = [  # This is the schema of nyc-tlc:yellow.trips
+            "vendor_id",
+            "pickup_datetime",
+            "dropoff_datetime",
+            "pickup_longitude",
+            "pickup_latitude",
+            "dropoff_longitude",
+            "dropoff_latitude",
+            "rate_code",
+            "passenger_count",
+            "trip_distance",
+            "payment_type",
+            "fare_amount",
+            "extra",
+            "mta_tax",
+            "imp_surcharge",
+            "tip_amount",
+            "tolls_amount",
+            "total_amount",
+            "store_and_fwd_flag",
+            "average_speed"
+        ]
+
+    def dict_to_csv(self, dictionary):
+        """This funciton converts a python dictionary to a CSV line. Note keys in output schema that
+        are missing in the dictionary or that contains commas (which happens occasionally with the
+        store_and_fwd_flags fields due to a data quality issue) will result in empty values.
+        Arguments:
+            dictionary: A dictionary containing the data of interest.
+        """
+        csv = ','.join([str(dictionary[key]) if dictionary.get(key) is not None else ''
+                        for key in self.output_schema])
+        return csv
+
+    def enhance_with_avg_speed(self, record):
+        """
+        This is the business logic for the average speed column to calculate for each record.
+        The desired units are miles per hour.
+        Arguments:
+            record: A dict record from the nyc-tlc:yellow Public BigQuery table to be transformed
+                    with an average_speed field. (This argument object gets modified in place by
+                    this method).
+        """
+        # Check if store_and_fwd_flag is a legal value 'Y' or 'N'
```

Comment:
Removed line 75. kept comment below calling out data quality issue to avoid confusion as to why we are doing this at all. Note this is necessary because this field sometimes is set to a comma which pass these records through the spark job writing a row with too many commas and will then trip up the load to BigQuery later in the DAG.